unittests/test_utils.hpp

```diff
@@ -70,13 +74,42 @@ auto push_input_trx(eosio::chain::controller& control, account_name account, sig
       trx.sign(testing::tester::get_private_key(account, "active"), control.get_chain_id());
    }
    auto ptrx = std::make_shared<packed_transaction>( trx, packed_transaction::compression_type::zlib );
-   auto fut = transaction_metadata::start_recover_keys( ptrx, control.get_thread_pool(), control.get_chain_id(), fc::microseconds::maximum(), transaction_metadata::trx_type::input );
-   auto r = control.push_transaction( fut.get(), fc::time_point::maximum(), fc::microseconds::maximum(), 0, false, 0 );
-   return r;
+
+   std::promise<transaction_trace_ptr> trx_promise;
+   std::future<transaction_trace_ptr> trx_future = trx_promise.get_future();
+
+   app->executor().post( priority::low, exec_queue::read_write, [&ptrx, &app, &trx_promise]() {
+      app->get_method<plugin_interface::incoming::methods::transaction_async>()(ptrx,
+                                                                                false, // api_trx
+                                                                                transaction_metadata::trx_type::input, // trx_type
+                                                                                true, // return_failure_traces
+           [&trx_promise](const next_function_variant<transaction_trace_ptr>& result) {
+              if( std::holds_alternative<fc::exception_ptr>( result ) ) {
+                 try {
+                    std::get<fc::exception_ptr>(result)->dynamic_rethrow_exception();
+                 } catch(...) {
+                    trx_promise.set_exception(std::current_exception());
+                 }
+              } else if ( std::get<chain::transaction_trace_ptr>( result )->except ) {
+                 try {
+                    std::get<chain::transaction_trace_ptr>(result)->except->dynamic_rethrow_exception();
+                 } catch(...) {
+                    trx_promise.set_exception(std::current_exception());
+                 }
+              } else {
+                 trx_promise.set_value(std::get<chain::transaction_trace_ptr>(result));
+              }
+           });
+   });
+
+   if (trx_future.wait_for(std::chrono::seconds(5)) == std::future_status::timeout)
+      throw std::runtime_error("failed to execute trx: " + ptrx->get_transaction().actions.at(0).name.to_string() + " to account: " + account.to_string());
```

Comment:
So we posted on the app_thread a lambda containing references to local variables. And, upon a `5s` timer, we can throw an exception and exit the stack frame holding these local variable. I Think it could be an issue if the app thread executes the lambda afterwards.
Maybe it would be better to `make_shared` the promise and pass a `shared_ptr` to it?

Reply:
Done.